### ISSUE TITLE
Document native plugin development process in Deno

### ISSUE THREAD

**Issue by ghost** [denoland/deno#4964]
I am currently trying to make use of Deno's native plugin API, and I've been having a hard time trying to gather the basic resources for plugin development. For the record, I've taken a look at a few implementations alongside the associated files with [`lib.rs`](https://github.com/denoland/deno/blob/master/core/lib.rs) found in `libdeno`.

Is there any documentation for plugin development in Deno? If not, can we look into making one? If there does exist a solution, I'd be willing to migrate from the horrors of Node Addons. What I'm primary looking for is a:

- Smooth development process and fluid production build release
- Method to compile plugin with TypeScript as single executable

Cf. [`deno-plugin-prepare`](https://github.com/manyuanrong/deno-plugin-prepare), [`deno-webview`](https://github.com/eliassjogreen/deno_webview/).

**Comment by chiefbiiko** [comment]
[create-deno-plugin](https://github.com/chiefbiiko/create-deno-plugin)

_**installation**_

``` sh
deno install \
  --allow-env \
  --allow-read \
  --allow-write \
  --dir=/usr/local/bin \
  create-deno-plugin \
  https://denopkg.com/chiefbiiko/create-deno-plugin@v0.2.0/main.ts
```
_**usage**_

``` sh
create-deno-plugin ./my-plugin
```

then `cd ./my-plugin` and `cargo build`

**Comment by undefinedbuddy** [comment]
Wow, it seems like `create-deno-plugin` can scaffold a Deno plugin exactly how I needed it. Does `rusty_v8` get built with the plugin (as part of a transitive dependency by `deno_core`)?

I'd really make use of some docs or even some more examples for guidance. cc @ry @piscisaureus

**Comment by chiefbiiko** [comment]
@undefinedbuddy 
yes, `rusty_v8` is part of the build
also, if ur wondering about plugin distribution, have a look at
[this release workflow](https://github.com/chiefbiiko/deno-libhydrogen/blob/v0.2.1/.github/workflows/release.yml) prebuilding for the big3 and
[this little script](https://github.com/chiefbiiko/deno-libhydrogen/blob/v0.2.1/lib.ts) pulling in those prebuilts for library consumers

**Comment by undefinedbuddy** [comment]
@chiefbiiko Thanks! This really helped a lot.

**Comment by undefinedbuddy** [comment]
I updated to v1.0.0-rc and it seems like `Deno.Plugin` has been removed from the global `Deno` namespace, alongside `openPlugin`. Is there anywhere I can track new API changes besides the changelog for this version? Any migration guide?

**Comment by ry** [comment]
@undefinedbuddy use `--unstable` flag to use plugins. Documented here https://doc.deno.land/https/raw.githubusercontent.com/denoland/deno/master/cli/js/lib.deno.unstable.d.ts

**Comment by undefinedbuddy** [comment]
Great! Does that mean in my type definitions I can just use `Deno.Plugin`, as in, is it available in the global `Deno` namespace after running with `--unstable`? Additionally, I lost track of the previous version I had installed and with this new update (v1.0.0-rc), I'm now experiencing errors in using certain APIs from `std/fs`.

**Comment by lu4** [comment]
> I am currently trying to make use of Deno's native plugin API, and I've been having a hard time trying to gather the basic resources for plugin development. For the record, I've taken a look at a few implementations alongside the associated files with [`lib.rs`](https://github.com/denoland/deno/blob/master/core/lib.rs) found in `libdeno`.
> 
> Is there any documentation for plugin development in Deno? If not, can we look into making one? If there does exist a solution, I'd be willing to migrate from the horrors of Node Addons. What I'm primary looking for is a:
> 
> * Smooth development process and fluid production build release
> * Method to compile plugin with TypeScript as single executable
> 
> Cf. [`deno-plugin-prepare`](https://github.com/manyuanrong/deno-plugin-prepare), [`deno-webview`](https://github.com/eliassjogreen/deno_webview/).

@undefinedbuddy I've also faced same issues, spent whole weekend trying to figure out how native plugin system works, I think I can also provide hand for making some docs on the process. Though from what I see now is plugin system not supported in full, https://doc.deno.land/https/raw.githubusercontent.com/denoland/deno/master/cli/js/lib.deno.unstable.d.ts refers only to `Deno.openPlugin`. Other constructs such as `Deno.core.ops` `Deno.core.dispatch` are not exposed yet if I'm not missing anything.

**Comment by shaunxu** [comment]
@lu4 I created `deno.core.d.ts` in my project. Content below

```typescript
declare namespace Deno {
    export const core: {
        ops(): { [key: string]: number };
        dispatch(opId: number, control: Uint8Array, zeroCopy?: Uint8Array): ArrayBufferView | ArrayBuffer | undefined;
        setAsyncHandler(opId: number, cb: (response: ArrayBufferView | ArrayBuffer | undefined) => void): void;
    }
}
```

And in the beginning of the `mod.ts` I added `/// <reference path="./deno.core.d.ts" />` and it works.

**Comment by lu4** [comment]
@shaunxu Thank you for provided `.d.ts` file, I was mentioning it to figure out a reason why the subject APIs were missing in deno itself, and whether the helping hand is necessary. But thank you for the effort!

**Comment by lu4** [comment]
For whom this might be useful: https://deno.land/manual/getting_started/typescript

**Comment by mknet** [comment]
@chiefbiiko Is your project `create-deno-plugin` still active? Or would you recommend using something like [deno-plugin-prepare](https://github.com/manyuanrong/deno-plugin-prepare)? 

@ry Is there any info you give me, if you guys plan any kind of doc how to implement a deno native plugin? Or is there already something I didn't find, yet?

Thanks for your help, guys!

**Comment by chiefbiiko** [comment]
@mknet i haven#t used `create-deno-plugin` with a recent `deno`
anways, it used to provide necessary and functioning boilerplate for intersecting `deno` and `rust`
i think `deno-plugin-prepare` helps in running native plugins but not *building* them
here is an example of a proper module making use of the native plugin feature: [deno-libhydrogen](https://github.com/chiefbiiko/deno-libhydrogen) (beware though its using the old op convention of a **control** and zero-copy buf - i think that api has changed)

**Comment by chiefbiiko** [comment]
released [`create-deno-plugin@v0.3.0`](https://github.com/chiefbiiko/create-deno-plugin/tree/master) which runs with the current `deno v1.5.2`

**Comment by bartlomieju** [comment]
With native plugins removal in #11152 this is no longer relevant
